## Centro Federal de Educação Tecnológica Celso Suckow da Fonseca  (CEFET-TJ)

## Programa de Pós Graduação em Ciência da Computação (PPCIC)

## Disciplina: Análise e Projetos de Algoritmos

## Professor: Diego Nunes Brandão

## Aluna: Cristiane Gea

## Resolução da 1ª Lista de Exercícios

**Dificuldades gerais identificadas**:

* Realizar as provas e demonstrações matemáticas.

* Entender os conceitos relacionados a Estruturas de Dados.

* Implementar a programação sem utilização de bibliotecas prontas.

* Conseguir interpretar corretamente o objetivo de cada enunciado.

### **Parte 1 - Análise de Complexidade**

**Contextualização**: Notação $O$ 

Sejam $f$ e  $g$ funções reais positivas de variável inteira $n$. Diz-se que $f$ é $O(g)$, escrevendo-se $f = O(g)$, quando existir uma constante $c > 0$ e um valor inteiro $n_0$, com $n \geq n_0$, tal que:
    
$$
f(n) \leq c.g(n)
$$
    
onde a função $g$ atua como um limite superior para valores assintóticos da função $f$.

**Contextualização**: Notação $\Theta$

Sejam $f$ e  $h$ funções reais positivas. Diz-se que $f$ é $\Theta(h)$, escrevendo $f = \Theta(h)$, quando ambas as condições $f = O(h)$ e $h = O(f)$ forem verificadas. A notação $\Theta$ exprime o fato de que as funções possuem a mesma ordem de grandeza assintótica.

**Observação**: Enquanto que a notação $O$ é utilizada para exprimir complexidades, a notação $\Theta$ é utilizada para exprimir limites superiores justos.

#### Questão 1

Explique o significado das seguintes expressões:

(a) $f(n)$ é $O(1)$

**Resposta**:

Um polinômio de grau $d$ é de ordem $O(n^d)$. Como uma constante pode ser considerada como um polinômio de grau 0, então uma constante é $O(n^0)$, ou seja, $O(1)$. Em outras palavras, a função possui complexidade constante.

Além disso, a função $f(n)$ está delimitado por cima por uma constante. Mais precisamente, há números positivos $c$ e $m$, tal que $f(n) \leq c$ para todo $n \leq m$.

(b) $f(n)$ é $\Theta(1)$

**Resposta**:

$f$ é assintoticamente constante se $f$ é assintoticamente equivalente a uma função constante, $f(n) \in \Theta(1)$ (que tem o mesmo significado de $f(n) = \Theta(1)$).
    
$f(n) = \Theta(1) = \Theta(n^0)$ designa um tempo de computação/complexidade constante. Em outras palavras, corresponde a uma situação (pouco usual) em que as instruções do programa são executadas um número fixo de vezes, independentemente do valor do input $n$.

Além disso, a função $f(n)$ está delimitada por cima e por baixo por uma constante. Para estabelecer isso formalmente é possível dizer que $f(n)$ é $O(1)$ e $f(n)$ é $\Omega(1)$. Uma definição precisa do último termo é que há números positivos $c$ e $m$, tal que $f(n) \geq c$ para todo $n \geq m$. Estes termos são utilizados para o tempo de execução de algoritmos, portanto, um limite inferior constante é uma declaração vazia. 

(c) $f(n)$ é $n^{O(1)}$

**Resposta**:

A função $f(n)$ é delimitada por cima por uma função polinomial de algum fixo, mas não especificado, ou mais precisamente há inteiros positivos $c$ e $m$, tal que $f(n) \leq n^c$ para todo $n \geq m$.

#### Questão 2

Assumindo que $f_1(n)$ é $O(g_1(n))$ e $f_2(n)$ é $O(g_2(n))$, prove as seguintes declarações: 

(a) $f_1(n) + f_2(n)$ é $O(\max (g_1(n), g_2(n)))$

**Resposta**

A definição convencional da notação $O$ para $f_1(n)$ e $f_2(n)$ são, respectivamente:

* $f_1(n) \leq c_1.g_1(n)$ para todo $n \geq n_1$, onde $c_1$ e $n_1$ existem e são números positivos.

* $f_2(n) \leq c_2.g_2(n)$ para todo $n \geq n_2$, onde $c_2$ e $n_2$ existem e são números positivos.

A partir das declarações acima, é possível escrever:

* $f_1(n) \leq c_1.\max (g_1(n), g_2(n))$ para todo $n \geq \max(n_1, n_2)$

* $f_2(n) \leq c_2.\max (g_1(n), g_2(n))$ para todo $n \geq \max(n_1, n_2)$

Adicionando os termos, para todo $n \geq \max(n_1, n_2)$:

$$
f_1(n) + f_2(n) \leq c_1.g_1(n) + c_2.g_2(n) \\
f_1(n) + f_2(n) \leq c_1.\max (g_1(n), g_2(n)) + c_2.\max (g_1(n), g_2(n)) \\
f_1(n) + f_2(n) \leq (c_1 + c_2).\max (g_1(n), g_2(n))
$$

Fazendo $c_3 = c_1 + c_2$ e $n_3 = \max(n_1, n_2)$, tem-se que para todo $n \geq n_3$:

$$
f_1(n) + f_2(n) \leq c_3.\max (g_1(n), g_2(n))
$$

Logo, por definição, $f_1(n) + f_2(n) = O(\max (g_1(n), g_2(n)))$.

(b) $f_1(n) * f_2(n)$ é $O(g_1(n) * g_2(n))$ (regra do produto)}

**Resposta**

A definição convencional da notação $O$ para $f_1(n)$ e $f_2(n)$ são, respectivamente:

* $f_1(n) \leq c_1.g_1(n)$ para todo $n \geq n_1$, onde $c_1$ e $n_1$ existem e são números positivos.

* $f_2(n) \leq c_2.g_2(n)$ para todo $n \geq n_2$, onde $c_2$ e $n_2$ existem e são números positivos.

A partir das declarações acima, é possível escrever:

* $f_1(n) \leq c_1.(g_1(n) * g_2(n))$ para todo $n \geq (n_1 * n_2)$

* $f_2(n) \leq c_2.(g_1(n) * g_2(n))$ para todo $n \geq (n_1 * n_2)$

Fazendo o produto entre os termos, para todo $n \geq (n_1 * n_2)$:

$$
f_1(n) * f_2(n) \leq c_1.(g_1(n) * g_2(n)) *  c_2.(g_1(n) * g_2(n))\\
f_1(n) * f_2(n) \leq (c_1 * c_2).(g_1(n) * g_2(n))
$$

Fazendo $c_3 = c_1 * c_2$ e $n_3 = n_1 * n_2$, tem-se que para todo $n \geq n_3$:

$$
f_1(n) * f_2(n) \leq c_3.(g_1(n) * g_2(n))
$$

Logo, por definição, $f_1(n) * f_2(n) = g_1(n) * g_2(n)$.

(c) $c$ é $O(1)$

**Resposta**

Pela regra da complexidade polinomial, é possível afirmar que se $P(n)$ é um polinômio de grau $k$, então $P(n) = O(n^k)$.

Considerando que $P(n) = f(n)$ e partindo da definição convencional da notação $O$, $f$ é $O(g)$ quando existir uma constante $d > 0$ e um valor inteiro $n_0$, com $n \geq n_0$, tal que:

$$
P(n) \leq d.n^k
$$

Para $k = 0$, é possível escrever que:

$$
P(n) \leq d.n^0 \\
P(n) \leq d
$$

Logo, por definição:

$$
P(n) = O(d)
$$

Pela regra da constante é possível escrever a expressão anterior como:

$$
P(n) = O(d.1) = d.O(1) = O(1)
$$

Como $P(n) = c$ (conforme enunciado), é possível escrever a expressão anterior como

$$
c = O(1)
$$

(d) $2^{2n + a}$ é $O(2^n)$

**Resposta**

A definição convencional da notação $O$ diz que, sejam $f$ e $g$ funções reais positivas de variáveis inteira $n$. Diz-se que $f$ é $O(g)$ quando existir uma constante $c > 0$ e um valor inteiro $n_0$, com $n \geq n_0$, tal que:
    
$$
f(n) \leq c.g(n)
$$

Sabe-se que $2^{2n + a} = 2^{2n}*2^a$ e que, eventualmente, uma constante múltiplo de $2^n$ é sempre maior que $2^{2n + a}$. Assim, multiplicando $2^n$ por $2^a$, o resultado obtido será maior ou igual a $2^{2n + a}$.

Considerando $c = 2$ e $n_0 = a$, tem-se que $2^{2n + a} \leq 2^{2n}*2^a$. Portanto, $2^{2n + a} = O(2^n)$.

#### Questão 3

Usando as mesmas hipóteses que o exercício anterior refute as seguintes declarações:

(a) $f_1(n) - f_2(n)$ é $O(g_1(n) - g_2(n))$

**Resposta**

A definição convencional da notação $O$ para $f_1(n)$ e $f_2(n)$ são, respectivamente:

* $f_1(n) \leq c_1.g_1(n)$ para todo $n \geq n_1$, onde $c_1$ e $n_1$ existem e são números positivos.

* $f_2(n) \leq c_2.g_2(n)$ para todo $n \geq n_2$, onde $c_2$ e $n_2$ existem e são números positivos.

A partir das declarações acima, é possível escrever:

* $f_1(n) \leq c_1.(g_1(n) - g_2(n))$ para todo $n \geq (n_1 - n_2)$

* $f_2(n) \leq c_2.(g_1(n) - g_2(n))$ para todo $n \geq (n_1 - n_2)$

Fazendo $f_1(n) - f_2(n)$, temos para todo $n \geq (n_1 - n_2)$:

$$
f_1(n) - f_2(n) \leq c_1.(g_1(n) - g_2(n)) -  c_2.(g_1(n) - g_2(n))\\
f_1(n) - f_2(n) \leq (c_1 - c_2).(g_1(n) - g_2(n))
$$

Fazendo $c_3 = c_1 - c_2$ e $n_3 = n_1 - n_2$, tem-se que para todo $n \geq n_3$:

$$
f_1(n) - f_2(n) \leq c_3.(g_1(n) - g_2(n))
$$

Diante disso, é preciso considerar três possibilidades:

* $f_1(n) > f_2(n)$

$$
f_1(n) > f_2(n) \Rightarrow f_1(n) - f_2(n) > 0\\
f_1(n) - f_2(n) > 0 \Rightarrow f_1(n) - f_2(n) \leq c_3.(g_1(n) - g_2(n))\\
f_1(n) - f_2(n) \leq c_3.(g_1(n) - g_2(n)) \Rightarrow f_1(n) - f_2(n) \textrm{ é } O(g_1(n) - g_2(n))
$$

* $f_1(n) < f_2(n)$

$$
f_1(n) < f_2(n) \Rightarrow f_1(n) - f_2(n) < 0\\
f_1(n) - f_2(n) < 0 \Rightarrow f_1(n) - f_2(n) \geq c_3.(g_1(n) - g_2(n))\\
f_1(n) - f_2(n) \geq c_3.(g_1(n) - g_2(n)) \Rightarrow f_1(n) - f_2(n) \textrm{ não é } O(g_1(n) - g_2(n))
$$

* $f_1(n) = f_2(n)$

$$
f_1(n) = f_2(n) \Rightarrow f_1(n) - f_2(n) = 0\\
f_1(n) - f_2(n) = 0 \Rightarrow 0 = c_3.(g_1(n) - g_2(n))\\
0 = c_3.(g_1(n) - g_2(n)) \Rightarrow f_1(n) - f_2(n) \textrm{ não é } O(g_1(n) - g_2(n))
$$

Considerando, como exemplo, que $f_1(n) = 2n^3 - 5n^2 + 7n$ e $f_2(n) = n^2 + 3n$. Por definição, é possível mostrar que:

* Se $f_1(n)$ é $O(g_1(n))$, então $f_1(n) \leq c_1.g_1(n)$. Logo, se $2n^3 - 5n^2 + 7n$ é $O(n^3)$, então $2n^3 - 5n^2 + 7n \leq c_1n^3$.

* Se $f_2(n)$ é $O(g_2(n))$, então $f_2(n) \leq c_2.g_2(n)$. Logo, se $n^2 + 3n$ é $O(n^2)$, então $n^2 + 3n \leq c_2n^2$.

Fazendo $f_1(n) - f_2(n)$, temos $2n^3 - 6n^2 + 4n$. Diante disso, $f_1(n) - f_2(n)$ é $O(n^3)$, diferente de $O(g_1(n) - g_2(n)) = O(n^3-n^2)$.

(b) $f_1(n)/f_2(n)$ é $O(g_1(n)/g_2(n))$

**Resposta**

A definição convencional da notação $O$ para $f_1(n)$ e $f_2(n)$ são, respectivamente:

* $f_1(n) \leq c_1.g_1(n)$ para todo $n \geq n_1$, onde $c_1$ e $n_1$ existem e são números positivos.

* $f_2(n) \leq c_2.g_2(n)$ para todo $n \geq n_2$, onde $c_2$ e $n_2$ existem e são números positivos.

A partir das declarações acima, é possível escrever:

* $f_1(n) \leq c_1. (g_1(n)/g_2(n))$ para todo $n \geq (n_1/n_2)$

* $f_2(n) \leq c_2.(g_1(n)/g_2(n))$ para todo $n \geq (n_1/n_2)$

Fazendo o qucient entre os termos, para todo $n \geq (n_1/n_2)$:

$$
\frac{f_1(n)}{f_2(n)} \leq \frac{c_1.(g_1(n)/g_2(n))}{c_2.(g_1(n)/g_2(n))}\\
\frac{f_1(n)}{f_2(n)} \leq \frac{c_1}{c_2}.\frac{(g_1(n)/g_2(n))}{(g_1(n)/g_2(n))}
$$

Fazendo $c_3 = (c_1/c_2)$ e $n_3 = (n_1/n_2)$, tem-se que para todo $n \geq n_3$:

$$
\frac{f_1(n)}{f_2(n)} \leq c_3.\frac{(g_1(n)/g_2(n))}{(g_1(n)/g_2(n))}
$$

$$
\textrm{Se } \frac{(g_1(n)/g_2(n))}{(g_1(n)/g_2(n))} = 1 \textrm{ , então } \frac{f_1(n)}{f_2(n)} \leq c_3\\
\frac{f_1(n)}{f_2(n)} \leq c_3 \Rightarrow \frac{f_1(n)}{f_2(n)} \textrm{ não é } O \left (\frac{g_1(n)}{g_2(n)}\right )
$$

Considerando, como exemplo, que $f_1(n) = n^3 + 2n + 1$ e $f_2(n) = n^2 + 1$. Por definição, é possível mostrar que:

* Se $f_1(n)$ é $O(g_1(n))$, então $f_1(n) \leq c_1.g_1(n)$. Logo, se $n^3 + 2n + 1$ é $O(n^2)$, então $n^2 + 2n + 1 \leq c_1n^3$.

* Se $f_2(n)$ é $O(g_2(n))$, então $f_2(n) \leq c_2.g_2(n)$. Logo, se $n^2 + 1$ é $O(n^2)$, então $n^2 + 1 \leq c_2n^2$.

Fazendo $f_1(n)/f_2(n)$ e considerando as propriedades assintóticas, temos

$$
\lim_{n\rightarrow \infty } \frac{n^3 + 2n + 1}{n^2 + 1} = \frac{\lim_{n\rightarrow \infty}(n^3 + 2n + 1)}{\lim_{n\rightarrow \infty}(n^2 + 1)} = \frac{\lim_{n\rightarrow \infty} n^3 (1 + 1/n^2 + 1/n^3)}{\lim_{n\rightarrow \infty} n^2 (1 + 1/n^2)} = \frac{\lim_{n\rightarrow \infty} n^3 (1 + 0 + 0)}{\lim_{n\rightarrow \infty} n^2 (1 + 0)} = \frac{\lim_{n\rightarrow \infty} n^3}{\lim_{n\rightarrow \infty} n^2} = \lim_{n\rightarrow \infty} n = \infty
$$

Logo, $f_1(n)/f_2(n)$ é diferente de $O(g_1(n)/g_2(n)) = c_1n^3/c_2n^2 = (c_1/c_2)(n^{3-2}) = (c_1/c_2)n$.

#### Questão 4

Escreva o algoritmo de busca binária (na forma recursiva e não recursiva) e faça a análise de tempo de execução do pior ca de cada algoritimo.

**Contextualização**

Diferentemente da busca sequencial, a busca binária começa examinando o item do meio. Segundo Celes et al. (2004), a ideia do algoritmo de busca binária é testar o elemento de busca com o valor do elemento que está armazenado no meio do vetor. Se o elemento de busca for igual ao elemento do meio, então a busca será encerrada pois o elemento foi localizado. Caso contrário, o elemento de busca é menor, ou maior, que o elemento do meio. Se o elemento de busca for menor que o elemento do meio, então estará na 1ª parte da lista; se for menor que o elemento do meio, estará na 2ª parte da lista.

Se o elemento de busca estiver em uma das partes da lista, o procedimento será repetido, considerando somente a parte restante. Por fim, o procedimento é continuamente repetido, subdividindo a parte de interesse, até encontrar o elemento ou chegar a uma parte da lista com tamanho zero. Diante disso, o tempo de pesquisa por um elemento em uma lista ordenada é reduzida, em comparação com a busca sequencial. Assim, a busca binária começa com um palpite de onde o elemento procurado pode estar. Inicialmente, o palpite é sempre escolher o elemento do meio da lista. 

O pior caso ocorre quando o elemento procurado não está localizado no vetor.

Por fim, na busca binária inicialmente todo o vetor é considerado para localizar o valor procurado. Contudo, a cada repetição, a parte considerada na busca é dividida à metade.

De acordo com Celes et al. (2016), o tamanho do vetor ao fim de cada repetição do laço do algoritmo:

Repetição | Tamanho do vetor
----------|----------
  0       | $n$
  1       | $n/2$
  2       | $n/4$
  3       | $n/8$
$\dots$   | $\dots$  
$\log n$  | 1

Logo, no pior caso $k$ repetições será igual a $\log n$.

**Resposta**

Dificuldades encontradas: (i) compreender as diferenças entre a busca binária recursiva e a busca binária iterativa e implementá-las na forma de código (achava que só havia um tipo de busca binária); (ii) realizar a análise de tempo de execução do pior caso no formato de como foi apresentado em aula (considerando o custo e o número de passos de cada etapa dos códigos).

Para avaliar o tempo de execução das funções desenvolvidas, foi utilizada uma lista na qual será realizada a busca binária.

In [1]:
# Definição do objeto onde ocorrerá a busca
a = list(range(10000000))

*Busca Binária Recurssiva*

Na versão recursiva, a entrada do algoritmo será uma lista de números inteiros, o valor procurado e o comprimento da lista. À cada iteração, o algoritmo exclui uma metade da lista e enalisa a outra metade. Como ponto de partida é realizado um palpite de que o valor procurado está no meio da lista. A saída do algoritmo será o valor booleano `True` (se o valor está na lista) ou `False` (se o valor não está na lista). 

Principais características do código:

* Assume que a lista está ordenada.

* Assume como palpite inicial o elemento contido na posição central da lista (elemento do meio ou mediana).

* Caso o elemento procurado não esteja na posição central, verifica se o elemento procurado é maior ou menor que o elemento central.

* Repete o processo até que o elemento procurado seja localizado na lista ou que seja verificado que ele não está contido na lista.

In [2]:
def bbRecursiva(lista, valor):
    # 1. Caso base: a lista está vazia
    if len(lista) == 0:
        return False
    else:
        # Palpite inicial: elemento do meio 
        mediana = len(lista)//2
        # 2. O palpite estava certo: o elemento está no meio da lista
        if lista[mediana] == valor:
            return True
        else:
            # 3. O palpite estava errado: os limites são atualizados 
            if valor < lista[mediana]:
                return bbRecursiva(lista[:mediana], valor)
            # item > lista[mediana]  
            else: 
                return bbRecursiva(lista[mediana+1:], valor)

Para simular o pior caso, o valor a ser procurado não estará dentro da lista.

In [3]:
%%time
bbRecursiva(a, 10000001)

Wall time: 203 ms


False

O algoritmo de busca binária recursiva levou 203 ms para localizar o nº 100001 dentro da lista.

Pseudocódigo para análise da complexidade de tempo de execução do pior caso:

```
def Busca_Binária_Recursiva (A, x):
    Se n = 0:                                            custo: C1 // nº de passos: 1
        retorna Falso                                    custo: C2 = 0
    Senão:                                               custo: C3 // nº de passos: 1
        mediana := n/2                                   custo: C4 // nº de passos: 1
        Se mediana = x                                   custo: C5 // nº de passos: 1
            Retorna Verdadeiro                           custo: C6 = 0
        Senão:                                           custo: C7 // nº de passos: 1
            Se x < mediana:                              custo: C8 // nº de passos: 1
                Devolve Busca_Binária_Recursiva (A, x)   custo: C9 // nº de passos: n/2
            Senão                                        custo: C10 // nº de passos: 1
                Devolve Busca_Binária_Recursiva (A, x)   custo: C11 // nº de passos: n/2
```

Cálculo do tempo de execução do algoritmo:

$$
T(n) = C_1 + C_3 + C_4 + C_5 + C_7 + C_8 + C_9 \times (n/2) + C_{10} + C_{11} \times (n/2) \\
T(n) = (C_9 \times (n/2) + C_{11} \times (n/2)) + C_1 + C_3 + C_4 + C_5 + C_7 + C_8 + C_{10} \\
T(n) = A(n/2) + B
$$

Considerando que o algoritmo encontre o valor procurado na primeira repetição, o tempo de execução corresponderia à $T(n) = A(n/2) + B$. Contudo, no pior caso o algoritmo não localizará o valor procurado. Neste caso, o algoritmo repetirá o processo de busca até que o tamanho do sub-vetor restante seja igual a 1 e o nº de repetições será igual a $\log n$. Portanto, considerando o pior caso, o tempo de execução do algoritmo corresponderá a $T(n) = A(\log n) + B$

O questionamento que surge é se é possível melhorar o algoritmo de busca binária. Embora não seja possível reduzir a complexidade assintótica do algoritmo, é possível tentar melhorá-lo de alguma forma. Uma possível melhoria do algoritmo é a implementação da sua versão iterativa. 

*Busca Binária Não Recursiva (Iterativa)*

Diferentemente da busca binária recursiva, na busca binária iterativa o algoritmo percorre a lista a cada iteração. Caso o palpite não esteja correto, os limites `first` e `last` são ajustados da parte da lista que está sendo analisada. A saída do algoritmo será o valor booleano `True` (se o valor está na lista) ou `False` (se o valor não está na lista). 

Obs.: enquanto que na versão recursiva, o espaço de busca do algoritmo é reduzido a cada chamada recursiva, na  versão iterativa, essa redução é realzada no espaço de busca a cada iteração do laço `while`.

Principais características do código:

* Considera que a lista está ordenada.

* Realiza a busca dentro de uma estrutura de repetição (representada por `while`), que é o diferencial em relação à busca recursiva.

* Assume como palpite inicial o elemento contido na posição central da lista (elemento do meio ou mediana).

* Caso o elemento procurado não esteja na posição central, verifica se o elemento procurado é maior ou menor que o elemento central.

* Repete o processo até que o elemento procurado seja localizado na lista ou que seja verificado que ele não está contido na lista.

In [4]:
def bbIterativa(lista, valor):
    # Definição do intervalo de busca // Verificação dos extremos da lista
    # A verificação dos extremos da lista é realizada no laço while (utilizado para percorrer a lista)
    inicio = 0
    fim = len(lista)-1
    resultado = False

    # Ajuste do espaço de busca
    while inicio <= fim and not resultado:
        # 1. Palpite inicial: elemento do meio 
        mediana = (inicio + fim)//2
        # 2. O palpite estava certo: o elemento está no meio da lista
        if lista[mediana] == valor:
            resultado = True
        # 3. O palpite não estava certo: ajuste dos limites first e last da parte da lista que está sendo analisada
        else:
            # Se o elemento procurado < elemento do meio, então o elemento procurado está na 1ª metade da sequência
            if valor < lista[mediana]:
                fim = mediana - 1
                # Se o elemento procurado > elemento do meio, então o elemento procurado está na 2ª metade da sequência 
            else:
                inicio = mediana + 1
    return resultado

Semelhante ao caso anterior, para simular o pior caso o valor a ser procurado não estará dentro da lista.

In [5]:
%%time
bbIterativa(a, 10000001)

Wall time: 0 ns


False

Em relação ao caso anterior, o algoritmo de busca binária iterativa levou um tempo muito inferior para realizar a busca pelo elemento procurado.

Pseudocódigo para análise da complexidade de tempo de execução do pior caso:

```
def Busca_Binária_Iterativa (A, x):               
    inicio := 0
    fim := n - 1
    enquanto inicio <= fim:                        custo: C1     // nº de passos: n
        mediana := [(inicio + fim)/2]              custo: C2 = 0 // nº de passos: 1
        se mediana == x:                           custo: C3 = 0 // nº de passos: 1
            retorna mediana                        custo: C4 = 0
        senão:                                     custo: C5 = 0 // nº de passos: 1
            se x < A[mediana]:                     custo: C6 = 0 // nº de passos: 1
                fim = mediana -1                   custo: C7     // nº de passos: n/2
            senão:                                 custo: C8 = 0 // nº de passos: 1
                inicio = mediana + 1               custo: C9     // nº de passos: n/2
    retorna resultado                              custo: C10 = 0
```

Cálculo do tempo de execução do algoritmo:

$$
T(n) = C_1 \times n + C_7 \times (n/2) + C_9 \times (n/2) \\
T(n) = (2C_1 + C_7 + C_9) \times  (n/2) \\
T(n) = A(n/2)
$$

Considerando que o algoritmo encontre o valor procurado na primeira repetição, o tempo de execução corresponderia à $T(n) = A(n/2)$. Contudo, no pior caso o algoritmo não localizará o valor procurado. Neste caso, o algoritmo repetirá o processo de busca até que o tamanho do sub-vetor restante seja igual a 1 e o nº de repetições será igual a $\log n$. Portanto, considerando o pior caso, o tempo de execução do algoritmo corresponderá a $T(n) = A(\log n)$.

Diante disso, é possível que o tempo de execução do algoritmo de busca binária recursiva é superior ao tempo de execução do algoritmo de busca binária iterativa.

Diferentemente do algoritmo de busca recursiva, o algoritmo de busca iterativa levou muito menos tempo para procurar o valor 100001 dentro da lista. Esse resultado é esperado, pois a função recursiva consome mais memória em relação à verão iterativa.

### **Parte 2 - Revisão de Programação Básica**

#### Questão 5

Faça um programa que leia um texto do usuário e conte o número de vogais que aparecem. O texto fornecido deve estar em um arquivo.

**Resposta**

A etapa inicial correspondeu à etapa de preparação, visto que contemplou a leitura do arquivo .txt e a separação dos caracteres. Por fim, foi realizada uma comparação entre os caracteres extraídos do texto com as vogais contidas em uma lista informada separadamente.

Limitação: não houve o tratamento para reconhecimento de vogais com acento (tanto na leitura do arquivo txt, como na função desenvolvida para contagem do nº de vogais).

Características principais do código:

* Leitura do texto contido no arquivo txt.

* Transformação do texto em uma lista de caracteres.

* Contagem do nº de vogais contidas no arquivo txt, com base em uma lista de vogais (considerando vogais minúsculas e maiúsculas).

In [6]:
# Leitura do arquivo no formato txt
texto = open(r"text_user.txt","r")
texto

<_io.TextIOWrapper name='text_user.txt' mode='r' encoding='cp1252'>

Após a leitura do arquivo .txt, é realizada a separação de cada caracter em uma lista para posterior análise de cada caracter.

In [7]:
# Criação de lista com caracteres do texto
caractere = [list(x) for x in texto]

# Exibição da lista de caracteres
caractere

[['O',
  's',
  ' ',
  'o',
  'u',
  't',
  'l',
  'i',
  'e',
  'r',
  's',
  ' ',
  's',
  'Ã',
  '£',
  'o',
  ' ',
  'o',
  'b',
  's',
  'e',
  'r',
  'v',
  'a',
  'Ã',
  '§',
  'Ã',
  'µ',
  'e',
  's',
  ' ',
  'q',
  'u',
  'e',
  ' ',
  'n',
  'Ã',
  '£',
  'o',
  ' ',
  's',
  'e',
  'g',
  'u',
  'e',
  'm',
  ' ',
  'u',
  'm',
  ' ',
  'c',
  'o',
  'm',
  'p',
  'o',
  'r',
  't',
  'a',
  'm',
  'e',
  'n',
  't',
  'o',
  ' ',
  'p',
  'a',
  'd',
  'r',
  'Ã',
  '£',
  'o',
  ' ',
  'e',
  'm',
  ' ',
  'u',
  'm',
  'a',
  ' ',
  's',
  'Ã',
  '©',
  'r',
  'i',
  'e',
  ' ',
  't',
  'e',
  'm',
  'p',
  'o',
  'r',
  'a',
  'l',
  '.',
  ' ']]

Elaboração de uma função para a verificação dos caracteres e identificação de quais são vogais (essa verificação é realizada por meio da comparação dos caracteres extraídos do arquivo com a lista de vogais).

In [8]:
# Função para contar o nº de vogais no texto
def contagem_vogais(texto):
    numero_vogais = 0
    vogais = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
    for x in texto:
        if x in vogais:
            numero_vogais = numero_vogais + 1
    return(numero_vogais)

In [9]:
# Contagem do nº de vogais no texto
contagem_vogais(str(caractere))

32

#### Questão 6

Escrever uma função (e um programa que execute tal função) que determine se uma matriz quadrada de dimensão $n(n<100)$ é uma matriz de permutação. Uma matriz quadrada é chamada de matriz de permutação se seus elementos são apenas $0$'s e $1$'s e se em cada linha e coluna da matriz existe um único valor 1.

Exemplo: A matriz $\begin{pmatrix}
1 & 0 & 0\\ 
0 & 0 & 1\\ 
0 & 1 & 0
\end{pmatrix}$ é uma matriz de permutação.

**Resposta**

Conforme descrito no enunciado, uma matriz de permutação é uma matriz que possui como elementos somente os números 0 e 1. Portanto, obrigatoriamente a soma dos elementos de cada linha tem que ser igual a 1 e a soma dos elementos de cada coluna também tem que ser igual a 1.

A partir do conceito de matriz de permutação foi desenvolvida uma função que verifica se a soma dos itens de cada linha é igual a 1 e a soma dos itens de cada coluna é igual a 1. A função retornará um valor booleano, com `True` caso a matriz analisada seja de permutação e `False`, caso contrário.

Dificuldade encontrada: compreender como implementar em código os requisitos necessários para que uma matriz seja de permutação.

Principais caracaterísticas do código:

* Cálculo da soma dos elementos de cada linha e de cada coluna da matriz.

* Verificação do valor das somas realizadas.

In [10]:
# Definição da função para determinar se uma matriz é de permutação
def permutation_check(order):
    # Verifica se a soma dos elementos da linha = 1
    if all(sum(rows) == 1 for rows in order):
        # Verifica se a soma dos elementos da coluna = 1
        return all(sum(columns) == 1 for columns in zip(*order))
    return False

*Teste 1: Verificação de uma matriz de números aleatórios*

In [11]:
# Geração de matriz de números aleatórios
import random
from random import randint
import numpy as np

min_val = 0
max_val = 9
num_rows = 3
num_cols = num_rows
random_matrix = np.random.randint(min_val,max_val,(num_rows,num_cols))

In [12]:
# Matriz de números aleatórios
random_matrix

array([[8, 1, 2],
       [7, 5, 6],
       [8, 7, 7]])

In [13]:
# Verifica se é matriz de números aleatórios é de permutação
permutation_check(random_matrix)

False

*Teste 2: Verificação de uma matriz identidade*

In [14]:
# Matriz Identidade (que possui todas as características de uma matriz de permutação)
identity_matrix = [
                  [1, 0, 0],
                  [0, 1, 0],
                  [0, 0, 1],
                  ]

In [15]:
# Verifica se é matriz de números aleatórios é de permutação
permutation_check(identity_matrix)

True

#### Questão 7

Dado o polinômio $p(x) = a_0 + a_1x + \dots + a_nx^n$, isto é, os valores de $n$ e de $a_0, a_1, \dots, a_n$, determine os coeficientes reais da primeira derivada de $p(x), utilize os conceitos aprendidos de alocação dinâmica e de funções.

**Contextualização**:

A primeira derivada do polinômio $p(x) = a_0 + a_1x + \dots + a_nx^n$ é $p'(x) = a_1 + 2a_2x + \dots + na_nx^{n-1}$.

Se $p(x) = p_1(x) + p_2(x) + \dots + p_n(x)$, então $p'(x) = p'_1(x) + p'_2(x) + \dots + p'_n(x)$

**Resposta**:

Como exemplo, será utilizada a função $p(x) = 2x^3 + 4x^2 + 3x + 1$, cuja primeira derivada é representada por $p ' (x) = 6x^2 + 8x + 3$.

Dificuldade encontrada: saber como transformar o conceito matemático de derivada de uma função polinomial em código e como implementar este código.

Limitação: Para esta questão fiz 2 testes. Para o primeiro teste, utilizei uma função que calculasse a primeira derivada de uma função polinomial, com a intenção de extrair seus coeficientes. Contudo, não consegui fazer a extração dos coeficientes desta derivada. Para o segundo teste, fiquei diretamente no cálculo dos coeficientes da derivada a partir da lista de coeficientes da função polinomial. Neste segundo teste, acredito ter tido êxito para o cálculo somente dos coeficientes da primeira derivada. 

*Primeira tentativa*

Na primeira tentativa, foram desenvolvidas 3 funções: uma para a leitura do polinômio na forma de string, uma para preparar o polinômio para o cálculo da primeira derivada e uma para o cálculo da primeira derivada.

Apesar de retornar a primeira derivada do polinômio (como string), esta solução não é otimizada (além de retornar uma string em vez de um array com os coeficientes da primeira derivada). Uma outra limitação é que, para o cálculo correto da primeira derivada, obrigatoriamente o polinômio tem que estar na forma $p(x) = a_nx^n + \dots + a_1x + a_0$. Caso o polinônio esteja na forma $p(x) = a_0 + a_1x + \dots + a_nx^n$, a primeira derivada não será calculada corretamente.

Principais características do código:

* Leitura da equação na forma de string.

* Decomposição da equação, de acordo com a potência de cada termo.

* Cálculo da derivada, considerando as seguintes regras de derivação:

    (i) Se $f(x) = a$ ($a$ é constante), então $f'(x) = 0$;
    
    (ii) Se $f(x) = ax$ ($a$ é constante), então $f'(x) = a$;
    
    (iii) (Regra da potência) Se $f(x) = x^a$ ($a$ é constante), então $f'(x) = a·x^{a-1}$;
    
    (iv) (Regra da soma de funções) $[f (x) + g (x)]' = f'(x) + g'(x)$.

In [16]:
##### Primeiro Teste #####
"""
    Um polinômio em uma única variável pode ser representada como um array contendo os coeficientes.
    Por exemplo, p(x) = 2x^3 + 4x^2 + 3x + 1 pode ser expresso como [2, 4, 3, 1]
    Passos:
        (i) Análise do polinômio
        (ii) Obtenção dos componentes básicos da equação (coeficientes, potência de x)
        (iii) Obtenção da fórmula da derivada
        (iv) Retorno à equação da forma como a entrada foi dada
"""

import re

# Função para a leitura do polinômio
def read_equation(equation):
    terms = equation.split('+')
    equation = [re.split('x\^?', t) for t in terms]
    equation_list = []
    for x in equation:
        try:
            coefficient = int(x[0])
        except ValueError:
            coefficient = 1
        try:
            exponential = int(x[1])
        except ValueError:
            exponential = 1
        except IndexError:
            exponential = 0
        equation_list.append((coefficient, exponential))
    return equation_list

# Função para mapear o polinômio para o cálculo da primeira derivada
def write_equation(equation_list):
    def str_exponential(index):
        if index == 0:
            return ''
        elif index == 1:
            return 'x'
        else:
            return 'x^%d' % (index,)
        
    def str_coeff(coeff):
        return '' if coeff == 1 else str(coeff)
    str_terms = [(str_coeff(coeff) + str_exponential(index)) for coeff, index in equation_list]
    return "+".join(str_terms)

# Função para calcular a primeira derivada 
def derivative(equation):
    equation_list = read_equation(equation)
    derivative_list = [(index*coeff, index-1) for coeff, index in equation_list[:-1]]
    return write_equation(derivative_list)

In [17]:
# Polinômio escolhido
equation = "2x^3 + 4x^2 + 3x + 1"

# Cálculo da 1ª derivada do polinômio escolhido
derivative(equation)

'6x^2+8x+3'

*Segunda Tentativa*

Na segunda tentativa foi desenvolvida somente uma função, cujo imput é um array com os coeficientes do polinômio $p(x)$ e output é um array contendo os coeficientes da primeira derivada de $p(x)$.

Uma outra limitação é que, para o cálculo correto da primeira derivada, obrigatoriamente o array contendo os coeficientes do polinômio $p(x) = a_0 + a_1x + \dots + a_nx^n$ tem que seguir uma ordem específica, ou seja, o array tem que ser $[a_0, a_1, \dots, a_n]$. Caso contrário, os coeficientes da primeira derivada não são calculados corretamente.

Principais características do código:

* Leitura dos coeficientes dos termos da função polinomial, na forma de array.

* Cálculo da derivada, considerando as seguintes regras de derivação:

    (i) Se $f(x) = a$ ($a$ é constante), então $f'(x) = 0$;
    
    (ii) Se $f(x) = ax$ ($a$ é constante), então $f'(x) = a$;
    
    (iii) (Regra da potência) Se $f(x) = x^a$ ($a$ é constante), então $f'(x) = a·x^{a-1}$;
    
    (iv) (Regra da soma de funções) $[f (x) + g (x)]' = f'(x) + g'(x)$.

In [18]:
##### Teste 2 #####
# Função para calcular a primeira derivada
def derivative_equation(coefficients):
    # Os coeficientes da primeira derivada serão adicionados na lista abaixo
    derivative_coefficients = []
    # Coeficientes da derivada
    for i in range(0,len(coefficients)):
        derivative_coefficients.append(coefficients[i]*i)
    # Retorna os coeficientes da derivada
    return derivative_coefficients

In [19]:
# Polinômio escolhido: p(x) = 2x^3 + 4x^2 + 3x + 1 = 1 + 3x + 4x^2 + 2x^3
# Coeficientes deste polinômio: [1, 3, 4, 2]
derivative_equation([1, 3, 4, 2])

[0, 3, 8, 6]

### **Parte 3 - Programação com Estruturas de Dados**

**Questão 8**

Conjuntos podem ser representados pela lista de seus elementos. Supondo esta representação, escreva funções para as operações usais de conjuntos.

Neste caso, vocês criarão novos nós para compor a lista C.

(a) União ($C = A \cup B$) 

(b) Interseção ($C = A \cap B$)

(c) Diferença ($C = A - B$)

(d) Pertinência ($C = A \in  B$)

**Contextualização**:

Segundo Necaise (2011), uma estrutura encadeada é uma coleção de objetos (nós), com cada um contendo dados e pelo menos uma refência, ou link, para outro nó. Uma lista encadeada é uma estrutura encadeada em que os nós estão conectados em sequência para formar uma lista.

Uma lista encadeamento simples é uma lista encadeada onde cada nó contém um único campo de ligação e permite uma completa travessia do primeiro até o último nó. Além disso, o campo de ligação entre os nós é unidirecional (ou seja, os nós são encadeados em uma única direção).


**Reposta**:

Para o desenvolvimento desta questão foram definidas 2 estruturas (representadas pelas classes), uma para representar o nó da lista (classe `Node`) e a outra para representar a lista em si (classe `LinkedList`). Além disso, foram definidas funções (dentro da classe `LinkedList`) para o desenvolvimento dos itens da questão.

Limitações: a função para calcular a operação $C = A - B$ não exibe o resultado. Além disso, não consegui definir uma função que imprimisse a lista com o formato `node1 -> node2 -> ... -> NULL`. Por fim, a operação de união não desconsidera os elementos em duplicidade (para isso, é preciso utilizar a função `unique`).

Dificuldades encontradas: compreender o conceito de listas encadeadas, bem como sua construção e sua dinâmica; e (ii) implementar código para a criação de uma lista encadeada; e (iii) implementar código que realize operações entre duas listas encadeadas.

Princiáis características do código:

* Implementação de uma classe com funções que cria estruturas de nós.

* Implementação de uma classe com funções que cria uma lista encadeada vazia e com funções que permitem a manipulação de seus elementos.

* Implementação de funções voltada para manipulação de elementos de uma única lista encadeada e de funções que envolvem operações envolvendo 2 listas encadeadas.

* Implementação de funções acessórias para suporte das funções diretamente relacionadas aos objetivos do enunciado da questão.

In [20]:
# Definição da classe Node (estrutura para representar o nó da lista)
class Node:
    # Função de criação (construtor)
    def __init__(self, data, next=None):
        self.data = data
        self.next = next
        
    """
        Um nó é representado por uma estrutura que contém 2 campos: a informação armazenada e o ponteiro para o
        próximo elemento da lista. O último nó da lista encadeada armazena um ponteiro inválido, sinalizando que não
        existe um próximo nó.
    """

In [21]:
# Definição da classe LinkedList
class LinkedList:
    # Função de criação (construtor)
    # A função que cria uma lista vazia deve ter como valor de retorno uma lista sem nenhum elemento
    def __init__(self):
        self.head = None   # inicialmente não aponta para ninguém
        self.next = None   # inicialmente não aponta para ninguém
        
    """
        A função init aloca dinamicamente o espaço para armazenar o novo nó da lista, guarda a informação no novo nó e 
        faz este nó apontar para o nó que era o primeiro da lista. O primeiro nó da lista é, então, atualizado.
    """
 
    # Identifica a referência anterior do nó (acesso ao nó anterior)
    def previous_node(self, data):
        current = self.head
        while (current and current.next != data):
            current = current.next
        return current
    
    # Copia os valores associados aos elementos armazenados em uma lista
    def duplicate(self):
        copy = LinkedList()
        current = self.head
        while current:
            node = Node(current.data)
            copy.insert(node)
            current = current.next
        return copy
    
    """
        Para duplicar uma lista, o algoritmo primeiro cria uma nova lista (copy), varre a lista copiando cada elemento e,
        posteriormente, insere os elementos copiados na nova lista (copy).
    """
    
    # Insere novos elementos a uma lista (insere elementos no final da lista)
    def insert(self, data):
        if self.head is None:
            self.head = data
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = data
            
    """
        Para inserir novos elementos à lista, o algoritmo primeiro verifica se a lista está vazia. Caso esteja vazia,
        o algoritmo coloca o novo elemento no primeiro nó (head). Caso contrário, o algoritmo atravessa a lista até o
        último nó, insere o novo elemento no último nó e cria um link entre os 2 últimos nós.
    """
 
    # Remove os valores associados aos elementos armazenados em uma lista
    def remove(self, data):
        prev_node = self.previous_node(data)
        if prev_node is None:
            self.head = self.head.next
        else:
            prev_node.next = data.next
    
    """
        Para remover um nó da lista é preciso religar o nó antecessor com o nó sucessor, de modo a evitar que elementos
        fiquem inacessíveis pela quebra do encadeamento sequencial.
    """
 
    # Imprime os valores associados aos elementos armazenados em uma lista
    def display(self):
        current = self.head
        while current:
            print(current.data, end = ' ')
            current = current.next
            
    """
        Para imprimir os elementos de uma lista, o algoritmo atravessa uma lista exibindo cada elemento desta lista.
    """

    # Inverte a posição dos elementos de uma lista (utilizado na questão 10)
    def reverse(self):
        prev = None
        current = self.head
        while (current is not None):
            next = current.next
            current.next = prev
            prev = current
            current = next
        self.head = prev

    """
        Para inverter a posição dos elementos de uma lista, o algoritmo varre a lista trocando de lugar o elemento
        com seu elemento antecessor (em outras palavras, troca o nó com o nó anterior).
    """

    # Analisa se determinados elementos pertencem a uma lista
    def pertinence(self, llist):
        if self.search(llist):
            return True 
        return False
    
    """
        Para verificar se uma lista pertence à outra o algoritmo utiliza a função search (abaixo). Assim, atravessa a 2ª
        lista e verifica se todos os elementos da primeira lista estão contidos na 2ª lista.
    """

    # Verifica se determinado elemento está contido em uma lista
    def search(self, llist):
        current = self.head
        while  current and current.data != llist:
            current = current.next
        return current
    
    """
        Para realizar busca de um elemento em uma lista, o algoritmo primeiro atravessa esta lista procurando
        se há correspondência do elemento procurado com algum elemento da lista.
    """

    # Remove elementos em duplicidade em uma lista
    def unique(llist):
        current1 = llist.head
        while current1:
            current2 = current1.next
            data = current1.data
            while current2:
                temp = current2
                current2 = current2.next
                if temp.data == data:
                    llist.remove(temp)
            current1 = current1.next
            
    """
        Para remover elementos em duplicidade, o algoritmo percorre a lista e verifica se há alguma correspondência.
        Caso haja correspondência, então remove os elementos em duplicidade da lista original.
    """            

    # Realiza a união dos elementos de duas listas
    def union(self, llist):
        # Verifica se a primeira lista está vazia (caso esteja vazia, retorna com a cópia da segunda lista)
        if self.head is None:
            join = llist.duplicate()
            return join
        # Verifica se a segunda lista está vazia (caso esteja vazia, retorna com a cópia da segunda lista)
        if llist.head is None:
            join = self.duplicate()
            return join
        #Caso as 2 listas não esteja vazia, retorna com a cópia de ambas as listas em uma única lista
        join = self.duplicate()
        tail = join.head
        while tail.next is not None:
            tail = tail.next
        llistClone = llist.duplicate()
        tail.next = llistClone.head
        return join
    
    """
        Para realizar a união de 2 listas, o algoritmo primeiro verifica se as listas estão vazias. Caso uma delas
        esteja vazia, o mesmo retorna com a cópia da outra lista. No caso de ambas as listas não estiverem vazia,
        o algoritmo cria lista nova (join), atravessa a 1ª lista e adiciona seus elementos à nova lista (join),
        posteriormente atravessa a 2ª lista e adiciona seus elementos à nova lista (join).
    """
    
    # Realiza a interseção entre elementos de duas listas
    def intersection(self, llist):
        if (self.head is None or llist.head is None):
            return LinkedList()
        equal = LinkedList()
        current1 = self.head
        while current1:
            current2 = llist.head
            data = current1.data
            while current2:
                if current2.data == data:
                    node = Node(data)
                    equal.insert(node)
                    break
                current2 = current2.next
            current1 = current1.next
        return equal 

    """
        Para realizar a interseção de 2 listas, o algoritmo primeiro verifica se ambas as listas estão vazias.
        Caso uma delas esteja vazia, retorna com uma lista vazia. No caso de ambas as listas não estiverem vazia,
        o algoritmo cria lista nova (equal), atravessa a 1ª lista e olha todos os elementos da 2ª lista. Se o elemento
        da 1ª lista também estiver contido na 2ª lista, adiciona o elemento à nova lista (equal).
    """
    
    # Realiza a operação de diferença entre 2 listas
    def difference(self, llist):
        current1 = self.head
        current2 = llist.head
        diff = LinkedList()
        diff = self.duplicate()
        while current1:
            current1 = current1.next
            while current2:
                diff.remove(current2.data)
                current2 = current2.next
        return diff

    """
        Para realizar a diferenla entre 2 listas, o algoritmo primeiro cria lista nova (diff), atravessa a 1ª lista
        e copia os seus elementos para a nova lista. Posteriormente, atravessa a 2ª lista e olha todos os seus elementos.
        Se o elemento da 1ª lista também estiver contido na 2ª lista, então remove estes elementos da nova lista (diff).
    """

In [22]:
# Criação de instância da lista encadeada
list1 = LinkedList()

# Armazenamento dos dados na lista encadeada
list1.insert(Node(int(1)))
list1.insert(Node(int(2)))
list1.insert(Node(int(3)))
list1.insert(Node(int(4)))
list1.insert(Node(int(5)))
list1.insert(Node(int(6)))
list1.insert(Node(int(7)))

# Impressão dos elementos contidos na lista encadeada (para conferência)
list1.display()

1 2 3 4 5 6 7 

In [23]:
# Criação de instância da lista encadeada
list2 = LinkedList()

# Armazenamento dos dados na lista encadeada
list2.insert(Node(int(2)))
list2.insert(Node(int(4)))
list2.insert(Node(int(6)))
list2.insert(Node(int(8)))
list2.insert(Node(int(10)))
list2.insert(Node(int(12)))
list2.insert(Node(int(14)))

# Impressão dos elementos contidos na lista encadeada (para conferência)
list2.display()

2 4 6 8 10 12 14 

União entre 2 listas encadeadas ($C = A \cup B$)

In [24]:
list_union = LinkedList()
list_union = list1.union(list2)
list_union.display()

1 2 3 4 5 6 7 2 4 6 8 10 12 14 

Interseção entre 2 listas encadeadas ($C = A \cap B$)

In [25]:
list_intersection = LinkedList()
list_intersection = list1.intersection(list2)
list_intersection.display()

2 4 6 

Diferença entre 2 listas encadeadas ($C = A - B$)

(nesta etapa, o código executa sem exibição de mensagem de erro, mas não exibe o resultado)

In [26]:
list_difference = LinkedList()
list_difference = list1.difference(list2)
list_difference.display()

Relação de pertinência entre duas listas encadeadas ($C = A \in  B$)

In [27]:
list1.pertinence(list2)

False

**Questão 9**

Considere listas implementadas por encadeamento duplo, então pede-se para implementar funções que:

(a) Localize/Pesquise/Encontre (search) elementos

(b) Concatenar/Intercalar (Merge) duas listas

(c) Dividir uma lista em várias ($k$)

(d) Copiar uma lista

(e) Ordernar (sort) uma lista por ordem crescente/decrescente

**Contextualização**

Segundo Celes et al. (2016), nas listas duplamente encadeadas cada elemento tem um ponteiro para o próximo elemento e um ponteiro para o elemento anterior. Assim, dado um elemento, é possível acessar ambos os elemento adjacentes (antecessor e sucessor). Portanto, em uma lista duplamente encadeada, cada nó possui uma informação e 2 referências (uma para o nó antecessor e um para o nó sucessor).

**Resposta**

Para o desenvolvimento desta questão foram definidas 2 estruturas (representadas pelas classes), uma para representar o nó da lista (classe `Node`) e a outra para representar a lista em si (classe `DoubleLinkedList`). Além disso, foram definidas funções (dentro da classe `DoubleLinkedList`) para o desenvolvimento dos itens da questão.

Limitações: a função para dividir uma lista em várias só consegue dividir a lista em 2 partes iguais (obrigando a lista ter somente número par de elementos). Além disso, não consegui definir uma função que imprimisse a lista com o formato `node1 <=> node2 <=> ... <=> NULL`.

Dificuldades encontradas: compreender o conceito de listas duplamente encadeadas, bem como sua construção e sua dinâmica; e (ii) implementar código para a criação de uma lista duplamente encadeada (mais complexa que a de uma lista encadeada); (iii) implementar código que concatenasse duas listas duplamente encadeadas; e (iv) implementar código que dividisse uma lista duplamente encadeada em diversas sublistas.

Princiáis características do código:

* Implementação de uma classe com funções que cria estruturas de nós.

* Implementação de uma classe com funções que cria uma lista duplamente encadeada vazia e com funções que permitem a manipulação de seus elementos.

* Implementação de funções voltada para manipulação de elementos de uma única lista duplamente encadeada e de funções que envolvem a concatenação dos elementos contidos em 2 listas duplamente encadeadas.

* Implementação de funções acessórias para suporte das funções diretamente relacionadas aos objetivos do enunciado da questão.

In [28]:
# Definição da classe Node
class Node:  
    # Construtor
    def __init__(self,data):    
        self.data = data    
        self.previous = None    # inicialmente não aponta para ninguém
        self.next = None        # inicialmente não aponta para ninguém
        
    """
        Em uma lista duplamente encadeada, cada nó contém não somente o componente dado e um link para o próximo nó,
        mas também um link para o nó anterior
    """

In [29]:
# Criação de classe SearchList
class DoubleLinkedList:    
    # Construtor
    def __init__(self):    
        self.head = None    # início da lista
        self.tail = None    # fim da lista
            
    # Adiciona elemento à lista    
    def add_node(self, data):    
        # Cria um novo nó    
        newNode = Node(data)
        # Se a lista estiver vazia    
        if(self.head == None):    
            # Tanto head quanto tail apontarão para o novo nó (newNode)    
            self.head = self.tail = newNode    
            # Nó antecessor ao head apontará para None    
            self.head.previous = None    
            # Nó sucessor de tail apontará para None    
            self.tail.next = None   
        else:   # Se a lista não estiver vazia 
            # O novo nó (newNode) será adicionado depois de tail    
            self.tail.next = newNode    
            # O nó antecesso (newNode) apontará para tail   
            newNode.previous = self.tail    
            # O novo nó (newNode) se tornará o novo tail    
            self.tail = newNode    
            # Como é o último nó, o próximo apontará para None    
            self.tail.next = None    
    
    """
        Para inserir um novo nó em uma lista duplamente encadeada é preciso conectar o nó novo ao nó antecessor e ao nó
        sucessor. O algoritmo primeiro verifica se a lista está vazia. Caso esteja vazia, insere o elemento no primeiro
        nó. Caso contrário, o novo elemento será adicionado depois do último nó, tornando o novo último nó da lista e
        apontando para NULL.
    """
    
    # Busca por um determinado elemento da lista    
    def search(self, value):    
        i = 1;   
        flag = False    
        # Nó atual apontará para head    
        current = self.head    
        # Verifica se a lista está vazia    
        if(self.head == None):    
            print("Lista está vazia")    
            return   
        # Enquanto a lista não estiver vazia
        while(current != None):    
            # Compara o valor procurado com cada nó na lista    
            if(current.data == value):    
                flag = True    
                break    
            current = current.next    
            i = i + 1   
        if(flag):    
            print("O nó está presenta na lista na posição: " + str(i))    
        else:    
            print("O nó não está presente na lista")
            
    """
        Para verificar se determinado elemento está contido na lista, o algoritmo atravessa a lista comparando o
        elemento procurado com os elementos da lista.
    """
    
    # Imprime elementos da lista
    def display(self):
        # Define novo nó current que apontará para head
        current = self.head;
        if(self.head == None):
            print("Lista está vazia")
            return
        while(current != None):
            # Exibe cada nó incrementando o ponteiro
            print(current.data)
            current = current.next
        print()
        
    """
        Para imprimir os elementos de uma lista, o algoritmo primeiro verifica se a lista está vazia.
        Caso esteja vazia, retorna com mensagem indicando que está vazia. Caso contrário, atravessa a lista
        exibindo cada elemento desta lista.
    """
    
    # Concatena elementos de duas listas
    def merge(self, value):
        if self.tail and value.head:
            self.tail.next = value.head
            value.head.previous = self.tail
            return self
        
    """
        Para concatenar os elementos das listas, o algoritmo localiza o último nó da primeira lista e, posteriormente,
        insere os elementos da segunda lista após o último nó da primeira lista.
    """
    
    # Calcula o nº de elementos da lista
    def lenght(self):
        temp = self.head   # Inicializa temp
        count = 0          # Inicializa count
        # Loop enquanto o fim da lista não é alcançado
        while (temp):
            count += 1
            temp = temp.next
        return count

    """
        Para calcular o tamanho da lista, o algoritmo atravessa a lista contando quantos elementos ela possui
    """

    # Divide a lista (neste caso, divide somente em 2 partes iguais)
    def split(self, a, b):
        first = self.head
        second = first.next
        while (first is not None and second is not None and first.next is not None):
            self.MoveNode(a, first)
            self.MoveNode(b, second)
            first = first.next.next
            if first is None:
                break
            second = first.next
            
    """
        Para dividir uma lista em duas listas, o algoritmo verifica se a lista está vazia. Caso não esteja vazia, o
        algoritmo realiza a separação dos elementos, de modo que a metade dos elementos fique na primeira sublista e
        a outra metade dos elementos fique na outra sublista. Para isso, o algoritmo utiliza a função MoveNode para
        mover os elementos da lista original para uma das sublistas.
    """
    
    # Move elementos de posição dentro da lista
    def MoveNode(self, dest, node):
        newNode = Node(node.data)
        if dest.head is None:
            dest.head = newNode
        else:
            newNode.next = dest.head
            dest.head = newNode

    """
        Para mudar a posição dos elementos de uma lista, o algoritmo primeiro cria um novo nó e, posteriormente,
        verifica se a posição de destino está vazia. Caso esteja vazia, muda o nó de lugar. Caso contrário, coloca
        o nó na posição posterior.
    """

    # Cria uma lista nova e copia os elementos da lista original
    def clone(self):
        temp = self.head
        cloneList = DoubleLinkedList()
        while temp:
            cloneList.add_node(temp.data)
            temp = temp.next
        return cloneList

    """
        Para duplicar uma lista, o algoritmo primeiro cria uma nova lista (cloneList), varre a lista copiando cada elemento e,
        posteriormente, insere os elementos copiados na nova lista (cloneList).
    """
    
    # Ordena os elementos da lista em ordem crescente
    def sort(self):
        if (self.head == None):
            return
        else:
            current = self.head
            while (current.next != None):
                index = current.next
                while (index != None):
                    if (current.data > index.data):
                        temp = current.data
                        current.data = index.data
                        index.data = temp
                    index = index.next
                current = current.next
    
    # Ordena os elementos da lista em ordem decrescente
    def inverse(self):
        if (self.head == None):
            return
        else:
            current = self.head
            while (current.next != None):
                index = current.next
                while (index != None):
                    if (current.data < index.data):
                        temp = current.data
                        current.data = index.data
                        index.data = temp
                    index = index.next
                current = current.next
                
    """
        As funções de ordenação (sort e inverse) possuem funcionamento semelhante. Para ordenar os elementos de uma lista,
        o algoritmo primeiro verifica se a lista está vazia. Caso não esteja, compara os elementos e os reposiciona conforme
        seus respectivos valores
    """

In [30]:
# Criação de instância da lista encadeada
listA = DoubleLinkedList()   

# Armazenamento dos dados na lista encadeada    
listA.add_node(1)    
listA.add_node(5)   
listA.add_node(4)   
listA.add_node(2)    
listA.add_node(3)
listA.add_node(10)

# Impressão dos dados da lista encadeada
listA.display()

1
5
4
2
3
10



In [31]:
# Criação de instância da lista encadeada
listB = DoubleLinkedList()   

# Armazenamento dos dados na lista encadeada
listB.add_node(2)    
listB.add_node(4)   
listB.add_node(5)   
listB.add_node(8)    
listB.add_node(9)
listB.add_node(20)

# Impressão dos dados da lista encadeada
listB.display()

2
4
5
8
9
20



Localização de elementos na lista

In [32]:
# Busca pelo nó 4    
listA.search(4)    

O nó está presenta na lista na posição: 3


In [33]:
# Busca pelo elemento 9    
listA.search(9)

O nó não está presente na lista


Merge entre 2 listas

In [34]:
listMerge = DoubleLinkedList()
listMerge = listA.merge(listB)
listMerge.display()

1
5
4
2
3
10
2
4
5
8
9
20



Dividir da lista em várias ($k$) (nesta questão consegui somente dividir a lista encadeada em 2 sublistas iguais)

In [35]:
sublist1 = DoubleLinkedList()
sublist2 = DoubleLinkedList()
listB.split(sublist1, sublist2)

In [36]:
sublist1.display()

9
5
2



In [37]:
sublist2.display()

20
8
4



Cópia (clone) da lista

In [38]:
listClone = DoubleLinkedList()
listClone = listB.clone()
listClone.display()

2
4
5
8
9
20



Ordenação dos elementos da lista

In [39]:
# Criação de instância da lista encadeada
listC = DoubleLinkedList()   

# Armazenamento dos dados na lista encadeada    
listC.add_node(100)    
listC.add_node(0)   
listC.add_node(20)   
listC.add_node(500)    
listC.add_node(15)
listC.add_node(9)

# Impressão dos elementos da lista encadeada
listC.display()

100
0
20
500
15
9



In [40]:
# Elementos da lista em ordem crescente
listC.sort()  
listC.display()

0
9
15
20
100
500



In [41]:
# Elementos da lista em ordem decrescente
listC.inverse()  
listC.display()

500
100
20
15
9
0



**Questão 10**

Escreva uma função para uma lista simplesmente encadeada, que inverta a direção das ligações

**Resposta**

Na minha compreensão, uma função que inverta a direção das ligações é uma função que inverterá uma lista encadeada. Para este fim, foram utilizadas as classes `Node` e `LinkedList`, além da função `reverse` (dentro da classe `LinkedList`).

Visto que a função utilizada está contida na classe `LinkedList` (implementada na questão 8), as limitações, dificuldades encontradas e principais características do código estão listadas no desenvolvimento da questão 8.

In [42]:
# Criação de instância da lista encadeada
listReverse = LinkedList()

# Armazenamento dos dados na lista encadeada
listReverse.insert(Node(int(42)))
listReverse.insert(Node(int(33)))
listReverse.insert(Node(int(151)))
listReverse.insert(Node(int(400)))
listReverse.insert(Node(int(5)))
listReverse.insert(Node(int(9)))
listReverse.insert(Node(int(1)))

# Impressão dos elementos da lista encadeada
listReverse.display()

42 33 151 400 5 9 1 

In [43]:
# Inversão da posição dos elementos da lista
listReverse.reverse()

# Impressão dos elementos da lista encadeada invertida
listReverse.display()

1 9 5 400 151 33 42 

**Questão 11**

Considere que um estacionamento da Av. Maracanã é composto por uma única alameda que guarda até dez carros. Existe apenas uma entrada/saída no estacionamento, e esta extremidade da alameda dá acesso justamente à Av. Maracanã. Se chegar um cliente para retirar um carro que não seja o mais próximo da saída, todos os carros bloqueando seu caminho sairão do estacionamento. O carro do cliente será manobrado para fora do estacionamento, e os outros carros voltarão a ocupar a mesma sequência inicial. Escreva um programa que processe um grupo de linhas de entrada. Cada linha de entrada contém um ‘E’, de entrada, ou um ‘S’ de saída, e o número da placa do carro. Presume-se que os carros cheguem e partam na mesma ordem que entraram no estacionamento. O programa deve imprimir uma mensagem sempre que um carro chegar ou sair. Quando um carro chegar, a mensagem deve especificar se existe ou não vaga para o carro no estacionamento. Se não houver vaga, o carro partirá sem entrar no estacionamento. Quando um carro sair do estacionamento, a mensagem deverá incluir o número de vezes em que o carro foi manobrado para fora do estacionamento para permitir que os outros carros saíssem.

**Contextualização**:

De acordo com Celes et al. (2016), a ideia fundamental da pilha é que todo o acesso aos seus elementos é realizado pelo seu topo. Como cada novo elemento é inserido no topo, os elementos de uma pilha, portanto, são retirados na ordem inversa à ordem em que foram introduzidos (regra LIFO). Diante disso, há duas operações básica que devem ser implementadas em uma estrutura de pilha: a operação de empilhar um novo elemento, inserindo-o no topo (`push`), e a operação de desempilhar um elemento, removendo-o do topo (`pop`).

**Resposta**

Devido às limitações existentes com relação ao estacionamento (existência de apenas uma entrada/saída no estacionamento), foi criada uma pilha para os registros dos carros dentro do estacionamento. Considerando a estrutura de pilha (LIFO), só há acesso ao último carro que entrou no estacionamento. Dadas às associações, o estacionamento corresponde ao espaço da memória e os dados correspondem aos carros, representados pelas suas respectivas placas. Além disso, foram implementadas duas classes, uma para a pilha (onde são inseridos e removidos os elementos) e para a inserção dos carros, indentificados pelas, respectivas, cores e placas.

Limitação: não consegui incluir o número de vezes em que o carro foi manobrado para fora do estacionamento para permitir que os outros carros saíssem.

Dificuldades encontradas: (i) compreender que a estrutura de dados pedida no enunciado é de pilha, compreender o conceito de pilha e implementá-la em código; e (ii) integrar as duas classes implementadas (uma para a pilha e a outra para manipulação dos elementos da pilha).

Principais características do código:

* Implementação de uma classe com funções que cria uma pilha vazia e com funções que permitem a manipulação de seus elementos.

* Implementação de uma classe para inserção dos elementos da pilha como registros de carros.

In [44]:
# Implementação da classe Pilha
class Garage:
    # Inicia com uma pilha vazia
    def __init__(self):
        self.itens = []
        self.capacity = 10          # capacidade máxima da pilha (estacionamento)
  
    # Verifica se pilha está vazia // verifica se o estacionamento está vazio
    def is_empty(self):
        if self.itens == []:
            print("Estacionamento está vazio")
        else:
            print("Estacionamento não está vazio")
    
    # Verifica se a pilha está cheia // verifica se há vagas disponíveis no estacionamento
    def isfull(self):
        if len(self.itens) == self.capacity:
            print("Estacionamento Lotado")
        else:
            print("Estacionamento tem vaga")
    
    # Adiciona elemento no topo da pilha (topo é o final da lista) // chegada de um novo carro ao estacionamento
    def push(self, item):
        self.itens.append(item)
        print('PUSH %s' %item)
  
    # Remove elemento do topo (final da lista) // saída de um carro já estacionado 
    def pop(self, item):
        print('POP %s' %item)
        return self.itens.pop()
  
    # Obtém o elemento do topo (mas não remove)
    def peek(self):
        # Em Python, índice -1 retorna último elemento (topo)
        return self.itens[-1]
  
    # Retorna o número de elementos da pilha // verifica quantos carros estão dentro do estacionamento
    def size(self):
        return len(self.itens)
  
    # Imprime o conteúdo da pilha 
    def print_stack(self):
        print(self.itens)

In [45]:
# Implementação da classe para movimentaçaõ dos carros (os carros são identificados pela respectiva placa e cor)
class Car:
    def __init__(self, plate, color):
        self.plate = plate
        self.color = color
        self.move = 1
    
    def __str__(self):
        return "Car [Plate = " + self.plate + ", Color = " + self.color + "]"

In [46]:
# Verifica se o estacionamento está vazio
if __name__ == '__main__':
    G = Garage()
    G.is_empty()

Estacionamento está vazio


In [47]:
# Verifica se há vagas no estacionamento
if __name__ == '__main__':
    G.isfull()

Estacionamento tem vaga


In [48]:
# Entrada de novos carros
if __name__ == "__main__":
    C = Car("ABC-1234", "White")
    G.push(C)
    
    C = Car("DEF-5678", "Black")
    G.push(C)
    
    C = Car("GHI-9101", "Silver")
    G.push(C)

PUSH Car [Plate = ABC-1234, Color = White]
PUSH Car [Plate = DEF-5678, Color = Black]
PUSH Car [Plate = GHI-9101, Color = Silver]


In [49]:
# Verifica se ainda há vagas
if __name__ == '__main__':
    G.isfull()

Estacionamento tem vaga


In [50]:
# Verifica se o estacionamento está vazio
if __name__ == '__main__':
    G.is_empty()

Estacionamento não está vazio


In [51]:
# Entrada de novos carros
if __name__ == "__main__":
    C = Car("AAA-1111", "Red")
    G.push(C)
    
    C = Car("ZZZ-0000", "Purple")
    G.push(C)
    
    C = Car("KKK-9999", "Yellow")
    G.push(C)
    
    C = Car("CAC-0101", "Pink")
    G.push(C)
    
    C = Car("DDD-0001", "Green")
    G.push(C)
    
    C = Car("XXX-8959", "Gold")
    G.push(C)
    
    C = Car("LAT-5587", "Blue")
    G.push(C)

PUSH Car [Plate = AAA-1111, Color = Red]
PUSH Car [Plate = ZZZ-0000, Color = Purple]
PUSH Car [Plate = KKK-9999, Color = Yellow]
PUSH Car [Plate = CAC-0101, Color = Pink]
PUSH Car [Plate = DDD-0001, Color = Green]
PUSH Car [Plate = XXX-8959, Color = Gold]
PUSH Car [Plate = LAT-5587, Color = Blue]


In [52]:
# Verifica se ainda há vagas disponíveis
if __name__ == '__main__':
    G.isfull()

Estacionamento Lotado


In [53]:
# Saída de um carro 
if __name__ == '__main__':
    G.pop(C)

POP Car [Plate = LAT-5587, Color = Blue]


In [54]:
# Verifica se ainda há vagas disponíveis
if __name__ == '__main__':
    G.isfull()

Estacionamento tem vaga


**Questão 12**

Para um dado número inteiro $n>1$, o menor inteiro $d<1$ que divide $n$ é chamado de fator primo. É possível determinar a fatoração prima de $n$ achando-se o fator primo $d$ e substituindo $n$ pelo quociente $n/d$, repetindo esta operação até que $n$ seja igual a 1. Utilizando estrutura de pilha para auxiliá-lo na manipulação de dados, implemente em um programa que compute a fatoração prima de um número imprimindo os seus fatores em ordem decrescente. Por exemplo, para $n=3960$, deverá ser impresso $11*5*3*3*2*2*2$. Justifique a escolha do TAD utilizado.

**Contextualização**:

Segundo Necaise (2011), uma pilha é um tipo de container com acesso restrito que armazena uma coleção linear, onde os dados são tais que o último item inserido é o primeiro a ser removido. Assim, novos itens são adicionados ou itens existentes são removidos pelo topo (final da pilha).

Segundo Celes et al. (2016), a ideia central do tipo abstrato de dados (TAD) é encapsular de quem usa determinado tipo a forma concreta com que o tipo foi implementado. Assim, a forma com que o tipo foiefetivamente implementado passa a ser um detalhe de implementação, que não deve afetar o uso do tipo nos mais diversos contextos. Com isso, a implementação é desacoplada do uso, facilitando a manutenção e aumentando o potencial de reutilização do tipo criado. Neste sentido, Nacaise (2011) exemplifica algumas operações encapsuladas em um TAD:

* `Stack()`: cria uma pilha vazia.

* `isEmpty()`: retorna um valor booleano indicando se a pilha está vazia.

* `length()`: retorna o número de itens em uma pilha.

* `pop()`: remove o último elemento da pilha (caso a pilha não esteja vazia.

* `peek()`: retorna a referência do item que está no topo de uma pilha não vazia, sem removê-la.

* `push(i)`: insere o item `i` no topo da pilha.

**Resposta**

Para a resolução desta questão foram realizadas duas implementações, uma para a classe `Stack` (implementação de uma estrutura de pilha) e a uma para a função `factorization` (decomposição de determinado número inteiro em fatores primos). Conforme visto abaixo, na classe `Stack` consta as funções necessárias para inicialização e manipulação dos elementos de uma pilha. 

Dentro da lista de operações encapsuladas por um TAD, foram utilizadas `Stack()`, para criação de uma pilha vazia, e `push()`, para inserir os fatores primos derivados da decomposição de determinado número inteiro.

Limitação: o output final não ficou no formato solicitado pelo enunciado.

Dificuldade encontrada: fatorar um número inteiro considerando a estrutura de uma pilha.

Principais características do código:

* Implementação de uma classe com funções que cria uma pilha vazia e com funções que permitem a manipulação de seus elementos.

* Implementação de função para a decomposição de um inteiro em seus fatores primos.

* Integração da função de decomposição com a classe de pilha por meio das operações encapsuladas `Stack()` e `push()`.

In [55]:
# Implementação da classe Pilha
class Stack:
    # Inicia com uma pilha vazia
    def __init__(self):
        self.itens = []
  
    # Verifica se pilha está vazia
    def is_empty(self):
        return self.itens == []
  
    # Adiciona elemento no topo (topo é o final da lista)
    def push(self, item):
        self.itens.append(item)
        print('PUSH %s' %item)
  
    # Remove elemento do topo (final da lista)
    def pop(self):
        print('POP')
        return self.itens.pop()
  
    # Obtém o elemento do topo (mas não remove)
    def peek(self):
        # Em Python, índice -1 retorna último elemento (topo)
        return self.itens[-1]
  
    # Retorna o número de elementos da pilha
    def size(self):
        return len(self.itens)
  
    # Imprime pilha na tela
    def print_stack(self):
        print(self.itens)

In [56]:
# Função para fatoração por números primos
def factorization(n):
    prime_factors = Stack()
    d = 2
    while d * d <= n:
        while n % d == 0:
            n //= d
            prime_factors.push(d)
        d += 1
    if n > 1:  # para evitar 1 como um fator
        assert d <= n
        prime_factors.push(n)
    return prime_factors    

In [57]:
# Fatoração por números primos
n = 3960
factorization(n)

PUSH 2
PUSH 2
PUSH 2
PUSH 3
PUSH 3
PUSH 5
PUSH 11


**Questão 13**:

Implemente uma estrutura de fila onde cada item da fila consista em um número variável de inteiros.

**Contextualização**:

Segundo Celes et al. (2016), a ideia fundamental da fila é a possibilidade de inserir um novo elemento no final da fila e a retirada do elemento no início (FIFO). Além disso, as filas devem ser restritivas no sentido de que um elemento não pode passar na frente de seu antecessor.

**Resposta**

Para a resolução desta questão foi realizada a implementação da classe `Queue`, que contém funções necessárias para a criação de uma fila vazia e manipulação dos seus respectivos elementos. Posteriormente, após instanciar a classe `Queue`, para a criação de uma fila vazia, foram inseridos novos elementos a esta fila (via função `enqueue`). Contudo, em vez de serem inseridos números inteiros, foram inseridas listas contendo quantidades variadas de números inteiros.

Dificuldades encontradas: (i) interpretar o tipo de fila que o enunciado está pedindo; e (ii) compreender o conceito de fila e implementá-la em código

Principais características do código:

* Implementação de uma classe com funções que cria uma fila vazia e com funções que permitem a manipulação de seus elementos.

* Possibilidade de inserir listas com tamanhos variados de inteiros.

In [58]:
# Implementação da classe Fila
class Queue:
    # Inicia com uma fila vazia
    def __init__(self):
        self.itens = []

    # Verifica se fila está vazia
    def is_empty(self):
        return self.itens == []
  
    # Adiciona elemento no início da fila
    def enqueue(self, item):
        self.itens.insert(0, item)
        print('ENQUEUE %s' %item)

    # Remove elemento do final da fila
    def dequeue(self):
        print('DEQUEUE')
        return self.itens.pop()

    # Retorna o número de elementos da fila
    def size(self):
        return len(self.itens)
  
    # Imprime a fila na tela
    def print_queue(self):
        print(self.itens)

In [59]:
if __name__ == '__main__':
    # Instancia a criação de nova fila
    Q = Queue()
    
    # Adiciona novos elementos à fila
    Q.enqueue([1])
    Q.enqueue([2,3])
    Q.enqueue([4,5,6])
    Q.enqueue([7,8,9,10])
    Q.enqueue([11,12,13,14,15])

ENQUEUE [1]
ENQUEUE [2, 3]
ENQUEUE [4, 5, 6]
ENQUEUE [7, 8, 9, 10]
ENQUEUE [11, 12, 13, 14, 15]


In [60]:
# Imprime os elementos da fila
if __name__ == '__main__':
    Q.print_queue()

[[11, 12, 13, 14, 15], [7, 8, 9, 10], [4, 5, 6], [2, 3], [1]]


**Questão 14**:

Existem partes de sistemas operacionais que cuidam da ordem em que os programas devem ser executados. Por exemplo, em um sistema de computação de tempo-compartilhado ("*time-shared*") existe a necessidade manter um conjunto de processos em uma fila, esperando para serem executados.

Escreva um programa que seja capaz de ler uma série de solicitações para:

(a) Incluir novos processos na fila de processo.

(b) Retirar da fila o processo com o maior tempo de espera.

(c) Imprimir o conteúdo da liusta de processo em determinado momento.

**Contextualização**:

Segundo o [Wikipedia](https://pt.wikipedia.org/wiki/Tempo_compartilhado), um sistema operacional de tempo compartilhado foi desenvolvido para permitir a interatividade a custos pequenso, permitindo que muitos usuários compartilhem o computador simultaneamente.

**Resposta**

Para o desenvolvimento desta questão, inicialmente foi implementada a classe `Queue`, que contém funções para inicialização de uma fila vazia e manipulação dos elementos da fila.

Principais características do código:

* Implementação de uma classe com funções que cria uma fila vazia e com funções que permitem a manipulação de seus elementos.

* Implementação de funções que inclui e remove elementos na fila, além de imprimir estes elementos.

In [63]:
# Implementação da classe Fila
class TimeShared:
    # Inicia com uma fila vazia
    def __init__(self):
        self.itens = []

    # Verifica se fila está vazia
    def is_empty(self):
        return self.itens == []
  
    # Adiciona elemento no início da fila
    def insert_process (self, item):
        self.itens.insert(0, item)
        print('INSERT %s' %item)

    # Remove elemento do final da fila
    def remove_process(self):
        print('REMOVE')
        return self.itens.pop()

    # Retorna o número de elementos da fila
    def size(self):
        return len(self.itens)
  
    # Imprime a fila na tela
    def display(self):
        print(self.itens)

Inclusão de novos processos na fila de processo (o processo 1 foi o primeiro a ser incluído e o processo 10 foi o último processo a ser incluído).

In [64]:
if __name__ == '__main__':
    # Cria a fila de processos
    Q = TimeShared()
    
    # Adiciona novos elementos
    Q.insert_process('Processo 1')
    Q.insert_process('Processo 2')
    Q.insert_process('Processo 3')
    Q.insert_process('Processo 4')
    Q.insert_process('Processo 5')
    Q.insert_process('Processo 6')
    Q.insert_process('Processo 7')
    Q.insert_process('Processo 8')
    Q.insert_process('Processo 9')
    Q.insert_process('Processo 10')

INSERT Processo 1
INSERT Processo 2
INSERT Processo 3
INSERT Processo 4
INSERT Processo 5
INSERT Processo 6
INSERT Processo 7
INSERT Processo 8
INSERT Processo 9
INSERT Processo 10


Remoção do processo com maior tempo de espera (pelo conceito de fila, o processo com maior tempo de espera é o primeiro processo incluído na fila, sendo, portanto, o processo 1).

Na fila, o primeiro elemento a entrar é o último elemento a sair e, portanto, quanto mais próximo da base mais tempo ficará armazenado na estrutura. Diante disso, o elemento que está a mais tempo na fila é aquele que está ocupando a primeira posição (o 1º elemento que foi inserido).

In [65]:
# Remoção do primeiro elemento da fila (elemento que está a mais tempo na fila)
Q.remove_process()

REMOVE


'Processo 1'

Impressão do conteúdo da lista de processo em determinado momento (como exemplo, será impresso o conteúdo da lista após a remoção do processo com maior tempo de espera).

In [66]:
# Impressão dos elementos da fila
Q.display()

['Processo 10', 'Processo 9', 'Processo 8', 'Processo 7', 'Processo 6', 'Processo 5', 'Processo 4', 'Processo 3', 'Processo 2']


**Questão 15**

Um deque é um conjunto de itens a partir do qual podem ser eliminados e inseridos itens em ambas as extremidades. Chame as duas estremidades de um deque esq. e dir. Como um deque pode ser representado como um vetor em C/C++? Escreva quatro funções em C/C++, RemDir, RemEsq, InsDir, InsEsq, para remover e inserir elementos nas extremidades esquerda e direita de um deque. Certifique-se de que as funções funcionem corretamente para o deque vazio e detectam o estouro e o underflow (tentativa de remoção quando a fila está vazia). Quais as desvantagens dessa implementação com relação à implementação por encadeamento/alocação dinâmica?

**Contextualização**:

Segundo Goodrich et al (2013), um Deque (ou *Double-Ended Queue*) é uma estrutura de dados tipo fila que suporta a inserção e a remoção de elementos tanto no início como no fim da fila. Assim, esta estrutura possui dois inícios e dois finais.

**Resposta**:

Visto que um Deque permite a inserção e remoção de elementos em amabas as extremidades (início//direita e fim//esquerda), para esta questão foi implementada uma classe `Deque`, que contém funções para inicialização de uma fila vazia e manipulação dos respectivos elementos. Portanto, as operações no Deque são:

* `InsDir()`: adiciona itens no início do Deque, ou seja, insere elementos na extremidade direita.

* `InsEsq()`: adiciona itens no fim do Deque, ou seja, insere elementos na extremidade esquerda.

* `RemDir()`: remove itens no início do Deque, ou seja, remove elementos na extremidade direita

* `RemEsq()`: remove itens no fim do Deque, ou seja, remove os elementos na extreminada esquerda

Com relação à implementação da classe, a inicialização de uma fila vazia foi realizada em um objeto tipo array. Contudo, no Python um vetor e uma lista são estruturas muito similares. Além disso, os objetos listas/vetores no Python podem ser utilizados de forma dinâmica, que pode ser aumentada ou diminuída em tempo de execução a partir da inserção ou remoção de elementos. Segundo Necaise (2011), a diferença entre vetor e lista no Python é o número de operações e a possibilidade de alteração nos respectivos tamanhos (o tamanho do vetor não pode ser alterado depois de ser criado).

Por outro lado, na linguagem C essa característica não é transparente, mas precisa ser implementada pelo programador. Assim, na criação de uma variável composta como um vetor, a alocação de memória é feito em tempo de compilação, de modo que a estrutura tenha o tamanho fixo durante toda a execução do programa (não é possível aumentar ou diminuir o tamanho do vetor posteriormente). Para permitir esta flexibilidade é preciso utilizar mecanismos para alocação dinâmica de memória.

Segundo Celes et. al (2016), na linguagem C, quando um vetor é declarado é preciso informar a dimensão do vetor (nº máximo de elementos que poderá ser armazenado no espaço de memória reservado para o vetor). Além disso, é preciso informar o tipo dos valores que serão armazenados no vetor, pois em um vetor somente é possível armazenar valores de um mesmo tipo. Contudo, este pré-dimensionamento do vetor é um fator limitante. Por outro lado, a alocação dinâmica possibilita é um meio de requisitar espaços de memória em tempo de execução.

Limitação: a estrutura de vetor no Python não possui o mesmo comportamento que no C.

Dificuldades encontradas: (i) identificar a estrutura de fila solicitada no enunciado; (ii) compreender o conceito de Double-Ended Queue e implementá-lo em código; e (iii) compreender as desvantagens desta implementação a partir do conceito de vetor no C.

Principais características do código:

* Implementação de uma classe com funções que cria uma fila vazia e com funções que permitem a manipulação de seus elementos.

* Maior felxibilidade com relação à estrutura tradicional de fila, visto que permite a remoção e inserção de elementos por ambas as extremidades. 

In [67]:
# Implementação da classe Deque (Double-Ended queue)
class Deque:
    # Inicia com uma fila vazia
    def __init__(self):
        self.itens = []

    # Verifica se fila está vazia
    def is_empty(self):
        return self.itens == []
  
    # Adiciona elemento na direita da fila
    def InsDir(self, item):
        self.itens.append(item)
        print('InsDir %s' %item)

    # Adiciona elemento na esquerda da fila
    def InsEsq(self, item):
        self.itens.insert(0, item)
        print('InsEsq %s' %item)

    # Remove elemento na direita da fila
    def RemDir(self):
        print('RemDir')
        return self.itens.pop()

    # Remove elemento na esquerda da fila
    def RemEsq(self):
        print('RemEsq')
        return self.itens.pop(0)
  
    # Retorna o número de elementos da fila
    def size(self):
        return len(self.itens)
  
    # Imprime fila na tela
    def print_deque(self):
        print(self.itens)

In [68]:
if __name__ == '__main__':
    # Cria fila
    D = Deque()
    
    # Imprime fila
    D.print_deque()

[]


In [69]:
if __name__ == '__main__':
    # Adiciona elementos à direita (início da fila)
    D.InsDir(1)
    D.InsDir(3)
    D.InsDir(5)
    D.InsDir(7)
    D.InsDir(9)
    
    # Adiciona elementos à esquerda (fim da fila)
    D.InsEsq(2)
    D.InsEsq(4)
    D.InsEsq(6)
    D.InsEsq(8)
    D.InsEsq(10)
    
    # Imprime fila
    D.print_deque()

InsDir 1
InsDir 3
InsDir 5
InsDir 7
InsDir 9
InsEsq 2
InsEsq 4
InsEsq 6
InsEsq 8
InsEsq 10
[10, 8, 6, 4, 2, 1, 3, 5, 7, 9]


In [70]:
if __name__ == '__main__':
    # Remove primeiro elemento à direita
    D.RemDir()
    
    # Remove primeiro elemento à esquerda
    D.RemEsq()
    
    # Imprime a fila
    D.print_deque()
    
    # Verifica se a fila está vazia
    print(D.is_empty())

RemDir
RemEsq
[8, 6, 4, 2, 1, 3, 5, 7]
False


### Referências

* Celes, W., Cerqueira, R., Rangel, J.L. (2016). Introdução a Estrutura de Dados com técnicas de programação em C. LTC - 2ª edição.

* Drozdek, A. (2013). Data Structures and Algorithms in C++. Cengage Learning - 4ª edição.

* Goodrich, M.T., Tamassia, R. & Goldwasser, M.H. (2013). Data Structures and Algorithms in Python. Wiley.

* Levada, A.L.M. (2021). Algoritmos e Estrutruas de Dados em Python: Complexidade e programação orientada a objetos". Universidade Federal de São Carlos - Departamento de Computação.

* Necaise, R.D. (2011). Data Structures and Algorithms Using Python. John Wiley & Sons.

* Wikipedia. Tempo Compartilhado. Link: https://pt.wikipedia.org/wiki/Tempo_compartilhado Acesso em 31/07/2021.